File: backend/app/api/routers/goals.py
"""Goals router - endpoints for nutrition goal management."""

from __future__ import annotations

import uuid

from fastapi import APIRouter, Depends, HTTPException, status
from sqlalchemy.ext.asyncio import AsyncSession

from app.api.deps import get_current_device_id
from app.db.session import get_session
from app.schemas.goal import (
    GoalCalculateRequest,
    GoalCalculateResponse,
    GoalCreateCalculatedRequest,
    GoalCreateManualRequest,
    GoalResponse,
    GoalUpdateRequest,
)
from app.services.goal_calculation import calculate_full_goal
from app.services.goals import (
    create_calculated_goal,
    create_manual_goal,
    get_current_goal,
    get_goal,
    soft_delete_goal,
    update_goal,
)


router = APIRouter(prefix="/goals", tags=["goals"])


@router.post("/calculate", response_model=GoalCalculateResponse)
async def goals_calculate(
    body: GoalCalculateRequest,
    device_id: uuid.UUID = Depends(get_current_device_id),
) -> GoalCalculateResponse:
    """
    Calculate BMR/TDEE/targets without saving.

    Use this endpoint to preview calculated values before saving.
    """
    calc = calculate_full_goal(
        gender=body.gender,
        birth_date=body.birth_date,
        height_cm=body.height_cm,
        current_weight_kg=body.current_weight_kg,
        activity_level=body.activity_level,
        weight_goal_type=body.weight_goal_type,
        weight_change_pace=body.weight_change_pace,
    )

    return GoalCalculateResponse(
        bmr_kcal=calc.bmr_kcal,
        tdee_kcal=calc.tdee_kcal,
        daily_calories_kcal=calc.daily_calories_kcal,
        protein_percent=calc.protein_percent,
        carbs_percent=calc.carbs_percent,
        fat_percent=calc.fat_percent,
        protein_grams=calc.protein_grams,
        carbs_grams=calc.carbs_grams,
        fat_grams=calc.fat_grams,
        water_ml=calc.water_ml,
        healthy_weight_min_kg=calc.healthy_weight_min_kg,
        healthy_weight_max_kg=calc.healthy_weight_max_kg,
        current_bmi=calc.current_bmi,
        bmi_category=calc.bmi_category,
    )


@router.get("/current", response_model=GoalResponse | None)
async def goals_get_current(
    device_id: uuid.UUID = Depends(get_current_device_id),
    session: AsyncSession = Depends(get_session),
) -> GoalResponse | None:
    """Get the current active goal for this device."""
    goal = await get_current_goal(session, device_id=device_id)
    if goal is None:
        return None
    return _goal_to_response(goal)


@router.get("/{goal_id}", response_model=GoalResponse)
async def goals_get(
    goal_id: uuid.UUID,
    device_id: uuid.UUID = Depends(get_current_device_id),
    session: AsyncSession = Depends(get_session),
) -> GoalResponse:
    """Get a specific goal by ID."""
    goal = await get_goal(session, device_id=device_id, goal_id=goal_id)
    if goal is None:
        raise HTTPException(status_code=status.HTTP_404_NOT_FOUND, detail="Goal not found")
    return _goal_to_response(goal)


@router.post("/calculated", response_model=GoalResponse, status_code=status.HTTP_201_CREATED)
async def goals_create_calculated(
    body: GoalCreateCalculatedRequest,
    device_id: uuid.UUID = Depends(get_current_device_id),
    session: AsyncSession = Depends(get_session),
) -> GoalResponse:
    """
    Create a calculated goal from body metrics.

    This will soft-delete any existing goals for this device.
    """
    goal = await create_calculated_goal(session, device_id=device_id, data=body)
    return _goal_to_response(goal)


@router.post("/manual", response_model=GoalResponse, status_code=status.HTTP_201_CREATED)
async def goals_create_manual(
    body: GoalCreateManualRequest,
    device_id: uuid.UUID = Depends(get_current_device_id),
    session: AsyncSession = Depends(get_session),
) -> GoalResponse:
    """
    Create a manual goal with direct calorie/macro input.

    This will soft-delete any existing goals for this device.
    """
    goal = await create_manual_goal(session, device_id=device_id, data=body)
    return _goal_to_response(goal)


@router.patch("/{goal_id}", response_model=GoalResponse)
async def goals_update(
    goal_id: uuid.UUID,
    body: GoalUpdateRequest,
    device_id: uuid.UUID = Depends(get_current_device_id),
    session: AsyncSession = Depends(get_session),
) -> GoalResponse:
    """Update goal targets (macros and water)."""
    goal = await update_goal(
        session,
        device_id=device_id,
        goal_id=goal_id,
        daily_calories_kcal=body.daily_calories_kcal,
        protein_percent=body.protein_percent,
        carbs_percent=body.carbs_percent,
        fat_percent=body.fat_percent,
        water_ml=body.water_ml,
    )
    if goal is None:
        raise HTTPException(status_code=status.HTTP_404_NOT_FOUND, detail="Goal not found")
    return _goal_to_response(goal)


@router.delete("/{goal_id}", status_code=status.HTTP_204_NO_CONTENT, response_model=None)
async def goals_delete(
    goal_id: uuid.UUID,
    device_id: uuid.UUID = Depends(get_current_device_id),
    session: AsyncSession = Depends(get_session),
) -> None:
    """Delete a goal."""
    ok = await soft_delete_goal(session, device_id=device_id, goal_id=goal_id)
    if not ok:
        raise HTTPException(status_code=status.HTTP_404_NOT_FOUND, detail="Goal not found")


def _goal_to_response(goal) -> GoalResponse:
    """Convert UserGoal model to GoalResponse schema."""
    return GoalResponse(
        id=goal.id,
        goal_type=goal.goal_type,
        gender=goal.gender,
        birth_date=goal.birth_date,
        height_cm=goal.height_cm,
        current_weight_kg=goal.current_weight_kg,
        activity_level=goal.activity_level,
        weight_goal_type=goal.weight_goal_type,
        target_weight_kg=goal.target_weight_kg,
        weight_change_pace=goal.weight_change_pace,
        bmr_kcal=goal.bmr_kcal,
        tdee_kcal=goal.tdee_kcal,
        daily_calories_kcal=goal.daily_calories_kcal,
        protein_percent=goal.protein_percent,
        carbs_percent=goal.carbs_percent,
        fat_percent=goal.fat_percent,
        protein_grams=goal.protein_grams,
        carbs_grams=goal.carbs_grams,
        fat_grams=goal.fat_grams,
        water_ml=goal.water_ml,
        healthy_weight_min_kg=float(goal.healthy_weight_min_kg) if goal.healthy_weight_min_kg else None,
        healthy_weight_max_kg=float(goal.healthy_weight_max_kg) if goal.healthy_weight_max_kg else None,
        current_bmi=float(goal.current_bmi) if goal.current_bmi else None,
        bmi_category=goal.bmi_category,
        created_at=goal.created_at,
        updated_at=goal.updated_at,
    )

File: client/src/services/api/goals.ts
import {
  GoalCalculateRequest,
  GoalCalculateResponse,
  GoalCreateCalculatedRequest,
  GoalCreateManualRequest,
  UserGoal,
} from "@models/types";
import { apiFetch } from "./http";
import { parseNumeric } from "@services/utils/parsing";

// API response types (snake_case from backend)
// Note: Backend Decimal fields come as strings in JSON
type GoalApiResponse = {
  id: string;
  goal_type: string;
  gender?: string;
  birth_date?: string;
  height_cm?: string | number; // Decimal from backend
  current_weight_kg?: string | number; // Decimal from backend
  activity_level?: string;
  weight_goal_type?: string;
  target_weight_kg?: string | number; // Decimal from backend
  weight_change_pace?: string;
  bmr_kcal?: number;
  tdee_kcal?: number;
  daily_calories_kcal: number;
  protein_percent: number;
  carbs_percent: number;
  fat_percent: number;
  protein_grams: number;
  carbs_grams: number;
  fat_grams: number;
  water_ml: number;
  healthy_weight_min_kg?: number;
  healthy_weight_max_kg?: number;
  current_bmi?: number;
  bmi_category?: string;
  created_at: string;
  updated_at: string;
};

type GoalCalculateApiResponse = {
  bmr_kcal: number;
  tdee_kcal: number;
  daily_calories_kcal: number;
  protein_percent: number;
  carbs_percent: number;
  fat_percent: number;
  protein_grams: number;
  carbs_grams: number;
  fat_grams: number;
  water_ml: number;
  healthy_weight_min_kg: number;
  healthy_weight_max_kg: number;
  current_bmi: number;
  bmi_category: string;
};

// Transform API response to frontend type
const transformGoalResponse = (response: GoalApiResponse): UserGoal => ({
  id: response.id,
  goalType: response.goal_type as UserGoal["goalType"],
  gender: response.gender as UserGoal["gender"],
  birthDate: response.birth_date,
  heightCm: parseNumeric(response.height_cm),
  currentWeightKg: parseNumeric(response.current_weight_kg),
  activityLevel: response.activity_level as UserGoal["activityLevel"],
  weightGoalType: response.weight_goal_type as UserGoal["weightGoalType"],
  targetWeightKg: parseNumeric(response.target_weight_kg),
  weightChangePace: response.weight_change_pace as UserGoal["weightChangePace"],
  bmrKcal: response.bmr_kcal,
  tdeeKcal: response.tdee_kcal,
  dailyCaloriesKcal: response.daily_calories_kcal,
  proteinPercent: response.protein_percent,
  carbsPercent: response.carbs_percent,
  fatPercent: response.fat_percent,
  proteinGrams: response.protein_grams,
  carbsGrams: response.carbs_grams,
  fatGrams: response.fat_grams,
  waterMl: response.water_ml,
  healthyWeightMinKg: response.healthy_weight_min_kg,
  healthyWeightMaxKg: response.healthy_weight_max_kg,
  currentBmi: response.current_bmi,
  bmiCategory: response.bmi_category as UserGoal["bmiCategory"],
  createdAt: response.created_at,
  updatedAt: response.updated_at,
});

const transformCalculateResponse = (
  response: GoalCalculateApiResponse,
): GoalCalculateResponse => ({
  bmrKcal: response.bmr_kcal,
  tdeeKcal: response.tdee_kcal,
  dailyCaloriesKcal: response.daily_calories_kcal,
  proteinPercent: response.protein_percent,
  carbsPercent: response.carbs_percent,
  fatPercent: response.fat_percent,
  proteinGrams: response.protein_grams,
  carbsGrams: response.carbs_grams,
  fatGrams: response.fat_grams,
  waterMl: response.water_ml,
  healthyWeightMinKg: response.healthy_weight_min_kg,
  healthyWeightMaxKg: response.healthy_weight_max_kg,
  currentBmi: response.current_bmi,
  bmiCategory: response.bmi_category,
});

/**
 * Calculate goal preview (BMR/TDEE/targets) without saving.
 */
export const calculateGoal = async (
  request: GoalCalculateRequest,
): Promise<GoalCalculateResponse> => {
  const body = {
    gender: request.gender,
    birth_date: request.birthDate,
    height_cm: request.heightCm,
    current_weight_kg: request.currentWeightKg,
    activity_level: request.activityLevel,
    weight_goal_type: request.weightGoalType,
    target_weight_kg: request.targetWeightKg,
    weight_change_pace: request.weightChangePace,
  };

  const response = await apiFetch<GoalCalculateApiResponse>(
    "/v1/goals/calculate",
    {
      method: "POST",
      body,
    },
  );

  return transformCalculateResponse(response);
};

/**
 * Get the current active goal for this device.
 */
export const getCurrentGoal = async (): Promise<UserGoal | null> => {
  const response = await apiFetch<GoalApiResponse | null>("/v1/goals/current");
  return response ? transformGoalResponse(response) : null;
};

/**
 * Get a specific goal by ID.
 */
export const getGoal = async (goalId: string): Promise<UserGoal> => {
  const response = await apiFetch<GoalApiResponse>(`/v1/goals/${goalId}`);
  return transformGoalResponse(response);
};

/**
 * Create a calculated goal from body metrics.
 */
export const createCalculatedGoal = async (
  request: GoalCreateCalculatedRequest,
): Promise<UserGoal> => {
  const body = {
    id: request.id,
    gender: request.gender,
    birth_date: request.birthDate,
    height_cm: request.heightCm,
    current_weight_kg: request.currentWeightKg,
    activity_level: request.activityLevel,
    weight_goal_type: request.weightGoalType,
    target_weight_kg: request.targetWeightKg,
    weight_change_pace: request.weightChangePace,
    protein_percent: request.proteinPercent,
    carbs_percent: request.carbsPercent,
    fat_percent: request.fatPercent,
    water_ml: request.waterMl,
  };

  const response = await apiFetch<GoalApiResponse>("/v1/goals/calculated", {
    method: "POST",
    body,
  });

  return transformGoalResponse(response);
};

/**
 * Create a manual goal with direct calorie/macro input.
 */
export const createManualGoal = async (
  request: GoalCreateManualRequest,
): Promise<UserGoal> => {
  const body = {
    id: request.id,
    daily_calories_kcal: request.dailyCaloriesKcal,
    protein_percent: request.proteinPercent,
    carbs_percent: request.carbsPercent,
    fat_percent: request.fatPercent,
    water_ml: request.waterMl,
  };

  const response = await apiFetch<GoalApiResponse>("/v1/goals/manual", {
    method: "POST",
    body,
  });

  return transformGoalResponse(response);
};

/**
 * Update goal targets (macros and water).
 */
export const updateGoal = async (
  goalId: string,
  patch: {
    dailyCaloriesKcal?: number;
    proteinPercent?: number;
    carbsPercent?: number;
    fatPercent?: number;
    waterMl?: number;
  },
): Promise<UserGoal> => {
  const body = {
    daily_calories_kcal: patch.dailyCaloriesKcal,
    protein_percent: patch.proteinPercent,
    carbs_percent: patch.carbsPercent,
    fat_percent: patch.fatPercent,
    water_ml: patch.waterMl,
  };

  const response = await apiFetch<GoalApiResponse>(`/v1/goals/${goalId}`, {
    method: "PATCH",
    body,
  });

  return transformGoalResponse(response);
};

/**
 * Delete a goal.
 */
export const deleteGoal = async (goalId: string): Promise<void> => {
  await apiFetch<void>(`/v1/goals/${goalId}`, { method: "DELETE" });
};
